test/mpi/attr/attrerr.c

```diff
@@ -99,14 +99,34 @@ int test_communicators(void)
         printf("delete function return code was MPI_SUCCESS in put\n");
     }
 
-    /* Because the attribute delete function should fail, the attribute
-     * should *not be removed* */
     err = MPI_Attr_delete(dup_comm_world, key_1);
     if (err == MPI_SUCCESS) {
         errs++;
         printf("delete function return code was MPI_SUCCESS in delete\n");
     }
 
+    /* Because the attribute delete function should fail, the attribute
+     * may *not be removed* */
+    err = MPI_Attr_get(dup_comm_world, key_1, &value, &flag);
+    if (err) {
+        errs++;
+        printf("Error with get\n");
+
+    }
+#if !defined(USE_STRICT_MPI) && defined(MPICH)
+    if (!flag) {
+        errs++;
+        printf("delete function failed but the attribute was removed\n");
+    }
+#endif
+    if (!flag) {
+        err = MPI_Attr_put(dup_comm_world, key_1, (void *) (MPI_Aint) (2 * world_rank));
+        if (err) {
+            errs++;
+            printf("Error with second put\n");
+        }
+    }
+
```

Comment:
Quoting the text from MPI-4.1 specification:
> This function (`comm_delete_attr_fn`) is called by MPI_COMM_FREE, MPI_COMM_DISCONNECT, MPI_COMM_DELETE_ATTR, and MPI_COMM_SET_ATTR to do whatever is needed to remove an attribute.
> The function returns MPI_SUCCESS on success and an error code on failure (in which case MPI_COMM_FREE will fail).
>
I am not sure why it only mentions `MPI_COMM_FREE` will fail, but I believe it supports that `MPI_COMM_DELETE_ATTR` and `MPI_COMM_SET_ATTR` also should fail. By fail, I think it commonly will be interpreted as the operation of deleting the attribute is not successful, i.e. not removed.

However, I concede that the exact behavior under an failure case is not well defined. So I guess your interpretation here is acceptable.

Reply:
@hzhou Lucky those that mostly care about the latest standard. Take the time machine back to MPI-1.1, and you will see the exact same sentence. Therefore, this text only mentioning Comm (and not Datatype nor Win) is a remnant of the time only communicators had attributes.

Reply:
> However, I concede that the exact behavior under an failure case is not well defined.

And that's the root of the problem. If you leave things open to interpretation, implementations will vary wildly. I just remember a long time ago discussion we had elsewhere with other people about how to interpret lower/upper bounds of datatypes and its propagation to new datatypes.

Reply:
Undefined is not the same as leaving to open interpretation. Often it is undefined because it is not possible or sensible to define one way or another. Take the error case, the failure may prevent us from removing the attribute. Yeah, we could blindly remove the attribute by ignoring the error return, but if the error return is due to some runtime exception, it may mean users didn't finish their cleanup thus we are creating a memory leak by removing the reference (attribute). But it is possible that the failure does not affect the cleanup or that there is nothing there is to be done for the cleanup. So undefined only have one clear message -- try to *avoid*. :)

Reply:
I'm not sure whether you are acknowledging that the MPI spec is vague/deficient/incomplete, or you are defending it as it is now. In any case, we will never reach an agreement in this thing, we have "axiomatic" differences that are irreconcilable.